velox/exec/tests/HashJoinTest.cpp

```diff
@@ -8705,6 +8707,277 @@ DEBUG_ONLY_TEST_P(
+// (cache hit).
+TEST_F(HashTableCachingTest, concurrent) {
+  // Use a unique query ID for this test to ensure clean cache state.
+  const std::string queryId =
+      "hashTableCachingConcurrentTest_" +
+      std::to_string(
+          std::chrono::steady_clock::now().time_since_epoch().count());
+
+  // Create probe and build vectors with distinct column names.
+  std::vector<RowVectorPtr> probeVectors = makeBatches(10, [&](int32_t) {
+    return makeRowVector(
+        {"t_k", "t_v"},
+        {
+            makeFlatVector<int32_t>(100, [](auto row) { return row % 23; }),
+            makeFlatVector<int64_t>(100, [](auto row) { return row; }),
+        });
+  });
+
+  std::vector<RowVectorPtr> buildVectors = makeBatches(5, [&](int32_t) {
+    return makeRowVector(
+        {"u_k", "u_v"},
+        {
+            makeFlatVector<int32_t>(50, [](auto row) { return row % 31; }),
+            makeFlatVector<int64_t>(50, [](auto row) { return row * 10; }),
+        });
+  });
+
+  createDuckDbTable("t", probeVectors);
+  createDuckDbTable("u", buildVectors);
+
+  auto planNodeIdGenerator = std::make_shared<core::PlanNodeIdGenerator>();
+
+  // Build the plan using HashJoinNode::Builder to set useHashTableCache.
+  auto buildPlanNode =
+      PlanBuilder(planNodeIdGenerator).values(buildVectors).planNode();
+
+  auto probePlanNode =
+      PlanBuilder(planNodeIdGenerator).values(probeVectors).planNode();
+
+  // Create HashJoinNode with useHashTableCache = true.
+  auto outputType = ROW(
+      {"t_k", "t_v", "u_k", "u_v"}, {INTEGER(), BIGINT(), INTEGER(), BIGINT()});
+
+  auto joinNode =
+      core::HashJoinNode::Builder()
+          .id(planNodeIdGenerator->next())
+          .joinType(core::JoinType::kInner)
+          .nullAware(false)
+          .leftKeys(
+              {std::make_shared<core::FieldAccessTypedExpr>(INTEGER(), "t_k")})
+          .rightKeys(
+              {std::make_shared<core::FieldAccessTypedExpr>(INTEGER(), "u_k")})
+          .left(probePlanNode)
+          .right(buildPlanNode)
+          .outputType(outputType)
+          .useHashTableCache(true)
+          .build();
+  const auto joinNodeId = joinNode->id();
+
+  const int numDrivers = 3;
+
+  // Helper to run a task and return the completed task.
+  // Both tasks use the same queryId so they share the cache entry.
+  auto runTask = [&]() {
+    auto queryCtx = core::QueryCtx::create(
+        driverExecutor_.get(),
+        core::QueryConfig({}),
+        std::unordered_map<std::string, std::shared_ptr<config::ConfigBase>>{},
+        cache::AsyncDataCache::getInstance(),
+        nullptr,
+        nullptr,
+        queryId);
+
+    return AssertQueryBuilder(joinNode, duckDbQueryRunner_)
+        .queryCtx(queryCtx)
+        .maxDrivers(numDrivers)
+        .assertResults(
+            "SELECT t.t_k, t.t_v, u.u_k, u.u_v FROM t, u WHERE t.t_k = u.u_k");
+  };
+
+  // Run two tasks concurrently using async.
+  auto future1 = std::async(std::launch::async, runTask);
```

Comment:
Can we have a number of threads that continuously running for multiple iterations? Thanks!